x-pack/plugins/fleet/public/applications/fleet/hooks/use_url_pagination.ts

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { useCallback, useEffect, useMemo } from 'react';
+import { useHistory, useLocation } from 'react-router-dom';
+import { useUrlParams } from './use_url_params';
+import { PAGE_SIZE_OPTIONS, Pagination, usePagination } from './use_pagination';
+
+type SetUrlPagination = (pagination: Pagination) => void;
+interface UrlPagination {
+  pagination: Pagination;
+  setPagination: SetUrlPagination;
+  pageSizeOptions: number[];
+}
+
+type UrlPaginationParams = Partial<Pagination>;
+
+/**
+ * Uses URL params for pagination and also persists those to the URL as they are updated
+ */
+export const useUrlPagination = (): UrlPagination => {
+  const location = useLocation();
+  const history = useHistory();
+  const { urlParams, toUrlParams } = useUrlParams();
+  const urlPaginationParams = useMemo(() => {
+    return paginationFromUrlParams(urlParams);
+  }, [urlParams]);
+  const { pagination, pageSizeOptions, setPagination } = usePagination(urlPaginationParams);
+
+  const setUrlPagination = useCallback<SetUrlPagination>(
+    ({ pageSize, currentPage }) => {
+      history.push({
+        ...location,
+        search: toUrlParams({
+          ...urlParams,
+          currentPage,
+          pageSize,
+        }),
+      });
+    },
+    [history, location, toUrlParams, urlParams]
+  );
+
+  useEffect(() => {
+    setPagination((prevState) => {
+      return {
+        ...prevState,
+        ...paginationFromUrlParams(urlParams),
+      };
+    });
+  }, [setPagination, urlParams]);
+
+  return {
+    pagination,
+    setPagination: setUrlPagination,
+    pageSizeOptions,
+  };
+};
+
+const paginationFromUrlParams = (urlParams: UrlPaginationParams): Pagination => {
+  const pagination: Pagination = {
+    pageSize: 20,
+    currentPage: 1,
+  };
+
+  // Search params can appear multiple times in the URL, in which case the value for them,
+  // once parsed, would be an array. In these case, we take the last value defined
+  pagination.currentPage = Number(
+    (Array.isArray(urlParams.currentPage)
+      ? urlParams.currentPage[urlParams.currentPage.length - 1]
```

Comment:
I believe .pop() will give the same result. If so, I'd prefer it as it's less statements to read.
```suggestion
      ? urlParams.currentPage.pop()
```

Reply:
Wow - why did I not think of that 🤦 
Thanks @jfsiii . going to merge this in, and will adjust this based on your comments in next PR.

Reply:
@paul-tavares no worries, It Depends :tm:, right :)  .pop() does alter the array. I didn't see it used again so seems it works for now, but there are trade-offs. 

I really wish we could `urlParams.currentPage[-1]` like perl/python.

After writing this I went looking and found that we can't do `arr[-2]` but `arr.at(-2)` is coming eventually via [`.at() method`](https://github.com/tc39/proposal-relative-indexing-method). I don't see anything about it in any of the feature support sites but it _is_ stage 3 now, so 🤞